benchmark/bls.cpp

```diff
@@ -0,0 +1,417 @@
+
+// To benchmark host functions directly without CDT wrappers,
+// we need to contruct a eosio::chain::webassembly::interface object,
+// as host functions are implemented in eosio::chain::webassembly::interface
+struct interface_in_benchmark {
+   interface_in_benchmark() {
+      // prevent logging from messing up benchmark results display
+      fc::logger::get(DEFAULT_LOGGER).set_log_level(fc::log_level::off);
+
+      // create a chain
+      fc::temp_directory tempdir;
+      auto conf_genesis = tester::default_config( tempdir );
+      auto& cfg = conf_genesis.second.initial_configuration;
+      // configure large cpu usgaes so expensive BLS functions like pairing
+      // can run a reasonable number of times within a trasaction time
+      cfg.max_block_cpu_usage        = 500'000;
+      cfg.max_transaction_cpu_usage  = 480'000;
+      cfg.min_transaction_cpu_usage  = 1;
+      chain = std::make_unique<tester>(conf_genesis.first, conf_genesis.second);
+      chain->execute_setup_policy( setup_policy::full );
+
+      // get hold controller
+      control = chain->control.get();
+
+      // create account and deploy contract for a temp transaction
+      chain->create_accounts( {"payloadless"_n} );
+      chain->set_code( "payloadless"_n, test_contracts::payloadless_wasm() );
+      chain->set_abi( "payloadless"_n, test_contracts::payloadless_abi() );
+
+      // construct a signed transaction
+      fc::variant pretty_trx = fc::mutable_variant_object()
+         ("actions", fc::variants({
+            fc::mutable_variant_object()
+               ("account", name("payloadless"_n))
+               ("name", "doit")
+               ("authorization", fc::variants({
+                  fc::mutable_variant_object()
+                     ("actor", name("payloadless"_n))
+                     ("permission", name(config::active_name))
+               }))
+               ("data", fc::mutable_variant_object()
+               )
+            })
+        );
+      trx = std::make_unique<signed_transaction>();
+      abi_serializer::from_variant(pretty_trx, *trx, chain->get_resolver(), abi_serializer::create_yield_function( chain->abi_serializer_max_time ));
+      chain->set_transaction_headers(*trx);
+      trx->sign( chain->get_private_key( "payloadless"_n, "active" ), control->get_chain_id() );
+
+      // construct a packed transaction
+      ptrx = std::make_unique<packed_transaction>(*trx, eosio::chain::packed_transaction::compression_type::zlib);
+
+      // build transaction context from the packed transaction
+      timer = std::make_unique<platform_timer>();
+      trx_timer = std::make_unique<transaction_checktime_timer>(*timer);
+      trx_ctx = std::make_unique<transaction_context>(*control, *ptrx, ptrx->id(), std::move(*trx_timer));
+      trx_ctx->max_transaction_time_subjective = fc::microseconds::maximum();
+      trx_ctx->init_for_input_trx( ptrx->get_unprunable_size(), ptrx->get_prunable_size() );
+      trx_ctx->exec();
```

Comment:
I don't think the `exec()` is needed, and probably cleaner not to call it. All you want is a setup apply context.

Reply:
`exec()` is needed to generate action traces. `apply_context`'s constructor calls `transaction_context::get_action_trace()`. Its assert
```
EOS_ASSERT( 0 < action_ordinal && action_ordinal <= trace->action_traces.size()
```
will fail if `exec()` is not called.